### PR TITLE
events

### PR DESCRIPTION
**Pull request by bmichotte**
- Change *ANDROID_{SDK|NDK}_PATH* to check for ENV path if they are not in ~/.rubymotion
- Add `on_touch_end`, `on_touch_move` and `on_touch_cancel` in `MG::Scene`

### CODE REVIEW

File: src/layer.cpp [hunk 1]
```diff
@@ -214,17 +223,114 @@ scene_on_touch_begin(VALUE rcv, SEL sel)
     }
     block = rb_retain(block); // FIXME need release...
 
-    auto listener = cocos2d::EventListenerTouchOneByOne::create();
-    listener->onTouchBegan = [block](cocos2d::Touch *touch,
+    auto scene = SCENE(rcv);
+    if (!RTEST(scene->touch_listener)) {
```
Comment (lrz) [inline src/layer.cpp]: We should do this instead:

```
if (scene->touch_listener == NULL) {
```

The `RTEST` macro is reserved for Ruby objects, and here it's a C++ object pointer.

File: src/layer.cpp [hunk 2]
```diff
@@ -214,17 +223,114 @@ scene_on_touch_begin(VALUE rcv, SEL sel)
...
+        scene->touch_listener = cocos2d::EventListenerTouchOneByOne::create();
+    } else {
```
Comment (lrz) [inline src/layer.cpp]: Could you replace this with:

```
}
else {
```

to conform to the project's coding style :-)

File: src/layer.cpp [hunk 3]
```diff
@@ -214,17 +223,114 @@ scene_on_touch_begin(VALUE rcv, SEL sel)
...
+    }
+    scene->touch_listener->onTouchCancelled = [block](cocos2d::Touch *touch,
+        cocos2d::Event *event) -> bool {
+    VALUE touch_obj = rb_class_wrap_new((void *)touch,
+        rb_cTouch);
+    return RTEST(rb_block_call(block, 1, &touch_obj));
+    };
+
+    return scene_add_listener(rcv, scene->touch_listener);
+}
```
Comment (lrz) [inline src/layer.cpp]: I think we should refactor the content of these functions (`on_touch_{begin,end,move.cancel}`) before merging this patch, as they are almost identical. Creating a new function with the common code and calling it from the other functions.

File: src/layer.cpp [hunk 4]
```diff
@@ -22,6 +22,7 @@ class mc_Scene : public cocos2d::LayerColor {
 	cocos2d::Scene *scene;
 	VALUE obj;
 	SEL update_sel;
+    cocos2d::EventListenerTouchOneByOne *touch_listener;
 
     mc_Scene() {
 	obj = Qnil;
```
Comment (lrz) [inline src/layer.cpp]: We should probably set `touch_listener` to `NULL` here (in the `mc_Scene` constructor).

**Comment by bmichotte** [comment]
change done

**Comment by lrz** [comment]
Perfect! One last comment, with this refactor the `scene_remove_listener` function doesn't seem to be needed anymore, could we remove it and inline the code in `scene_on_touch_event `?

**Comment by lrz** [comment]
I added you as a collaborator, so feel free to merge this yourself when you're done -- thank you so much for the patch, I was actually talking with @amirrajan a few minutes ago about the lack of touch events :-)

**Comment by bmichotte** [comment]
Thanks !

Even with collaborator rights, I'll continue to send PR !

**Comment by lrz** [comment]
\o/

**Pull request merged**